modyn/evaluator/internal/grpc/evaluator_grpc_servicer.py

```diff
@@ -125,31 +126,34 @@ def evaluate_model(self, request: EvaluateModelRequest, context: grpc.ServicerCo
                 evaluation_started=False, eval_aborted_reason=EvaluationAbortedReason.MODEL_NOT_EXIST_IN_STORAGE
             )
 
-        dataset_info = request.dataset_info
-        dataset_size_req = GetDatasetSizeRequest(
-            dataset_id=request.dataset_info.dataset_id,
-            start_timestamp=dataset_info.start_timestamp if dataset_info.HasField("start_timestamp") else None,
-            end_timestamp=dataset_info.end_timestamp if dataset_info.HasField("end_timestamp") else None,
-        )
-        dataset_size_response: GetDatasetSizeResponse = self._storage_stub.GetDatasetSize(dataset_size_req)
-
-        dataset_size = dataset_size_response.num_keys
-        if not dataset_size_response.success:
-            logger.error(
-                f"Total number of keys for dataset {dataset_size_req.dataset_id} cannot be fetched. "
-                f"Evaluation cannot be started."
-            )
-            return EvaluateModelResponse(
-                evaluation_started=False, eval_aborted_reason=EvaluationAbortedReason.DATASET_NOT_FOUND
+        data_sizes: list[int] = []
+        for interval in request.dataset_info.evaluation_intervals:
+            dataset_size_req = GetDatasetSizeRequest(
+                dataset_id=request.dataset_info.dataset_id,
+                start_timestamp=interval.start_timestamp if interval.HasField("start_timestamp") else None,
+                end_timestamp=interval.end_timestamp if interval.HasField("end_timestamp") else None,
             )
+            dataset_size_response: GetDatasetSizeResponse = self._storage_stub.GetDatasetSize(dataset_size_req)
 
-        if dataset_size == 0:
-            logger.info(
-                f"Dataset {dataset_size_req.dataset_id} is empty in given time interval. Evaluation cannot be started."
-            )
-            return EvaluateModelResponse(
-                evaluation_started=False, eval_aborted_reason=EvaluationAbortedReason.EMPTY_DATASET
-            )
+            dataset_size = dataset_size_response.num_keys
+            if not dataset_size_response.success:
+                logger.error(
+                    f"Total number of keys for dataset {dataset_size_req.dataset_id} cannot be fetched. "
+                    f"Evaluation cannot be started."
+                )
+                return EvaluateModelResponse(
+                    evaluation_started=False, eval_aborted_reason=EvaluationAbortedReason.DATASET_NOT_FOUND
+                )
+
+            if dataset_size == 0:
+                logger.info(
+                    f"Dataset {dataset_size_req.dataset_id} is empty in given time interval. "
+                    f"Evaluation cannot be started."
+                )
```

Comment:
What happens here with the return if only one out of several intervals is empty? I am a bit unsure whether we can safely return here.

Reply:
I guess we might need to have the eval_aborted_reason= / evaluation_started= per interval, since different intervals might fail/work

Reply:
Unfortunately, I have to disagree here. There are strategies that generate intervals in a fixed distance, e.g. [0,1000] and [1000,2000] etc. For example when running a regular CGLM pipeline, this will occur.

However, I am not sure we need to have a lot of engineering effort here. I see two solutions

1. Include the interval or some interval id that the supervisor generates in the metric reply. Then you immediately have a 1:1 mapping and it should be easy to pass-through

2. instead of failing here, return None / 0 for the metrics and just not start an evaluation. i am not sure if this has side effects (@robinholzi) but might be a quick fix.

However, I do think we need a 1:1 mapping since exceptions can occur. Exceptions sometimes happen (e.g., some random error while fetching data) and then the ordering may be disturbed I think (although it affects all intervals in the current batch of intervals that have not been processed). Not sure how you'd handle those right now?

Reply:
We definitely need this 1:1 mapping, often times, the test datasets don't have enough samples at the beginning yielding empty evaluations. With batched processing, we wouldn't evaluate anything for one model just because the first evaluation interval has no samples. The same occurs when using a Slicing / periodic strategy that generates smaller intervals (more likely to have no test data).
I think returning a list of EvaluationResults where every EvaluationResults uses (interval_start, interval_end) as unique identifier (for a certain model / dataset combination) is desirable. (@MaxiBoether's option 1)

Reply:
Thanks a lot, Robin and Maxi for providing extra insight! Then I guess I will use simply the interval_idx, i.e. the index in which the interval is present in the interval list, as the index. I think it is lighter as (interval_start, Interval_end).

For other transient failures during evaluation that manifest themselves as exceptions, currently when one happens, I just fail all requests in this batch, as I think there is something fundamentally wrong. but what do you suggest? @MaxiBoether

Reply:
It depends. If that is the easiest solution that's fine (we don't have a restart logic for failed evaluations anyways currently). Generally, what will happen is that at some point during the evaluation we see an exception and all intervals after the current ones also fail. The intervals for which we've already collected results we could theoretically still transfer the metrics. The most common failure I've seen (not very often, but actually the only one) is the connection between evaluator <-> storage just terminating for whatever reason, so something that did not really affect the numbers for the previous intervals that still worked. However, we can play it safe and just fail that batch

Reply:
Okay I made the logic dependent on the client: when an exception is met, is up to the client to decide whether to retrieve results prior to the exception. The server is ready to provide them